lib/PipelineC/PipelineCTracing.cpp

```diff
@@ -0,0 +1,283 @@
+    OS << "[";
+    for (uint64_t I = 0; I < Length; I++) {
+      OS << reprString(StringList[I]);
+      if (I < Length - 1) {
+        OS << ", ";
+      }
+    }
+    OS << "]\n";
+    OS.flush();
+  }
+
+  template<typename T>
+  // clang-format off
+  requires (isRPType<remove_constptr<remove_constptr<T>>>())
+  void printList(T PtrList, const uint64_t Length) {
+    // clang-format on
+    OS << "[";
+    for (uint64_t I = 0; I < Length; I++) {
+      OS << "P" << (void *) PtrList[I];
+      if (I < Length - 1) {
+        OS << ", ";
+      }
+    }
+    OS << "]\n";
+    OS.flush();
+  }
+
+  void printVoid() {
+    OS << "null\n";
+    OS.flush();
+  }
+
+  void endArguments() {
+    OutputtingArguments = false;
+    OS << "  return: ";
+    OS.flush();
+  }
+};
+
+static std::optional<PipelineCTracer>
+  Tracer = []() -> std::optional<PipelineCTracer> {
+  if (auto Path = llvm::sys::Process::GetEnv(TracingEnv)) {
+    return std::make_optional(*Path);
+  } else {
+    return std::nullopt;
+  }
+}();
+
+template<ConstexprString Name, int I, int N, typename... T>
+static void handleArg(std::tuple<T...> Args) {
+  Tracer->newArgument();
+  using ArgT = decltype(std::get<I>(Args));
+  using RArgT = std::remove_reference_t<ArgT>;
+  ArgT Arg = std::get<I>(Args);
+  constexpr int LH = LengthHint<Name, I>;
+  if constexpr (LH >= 0) {
+    // Handle arguments with length hints
+    if constexpr (std::is_same_v<RArgT, const char *>) {
+      // Buffer
+      Tracer->printBuffer(Arg, std::get<LH>(Args));
+    } else {
+      // Array-like
+      Tracer->printList<RArgT>(Arg, std::get<LH>(Args));
+    }
+  } else {
+    if constexpr (isDestroy<Name>()) {
+      // _destroy methods always take 1 argument and it's always a pointer
+      static_assert(N == 1);
+      Tracer->printPointer((void *) Arg);
+    } else {
+      Tracer->printValue<RArgT>(Arg);
+    }
+  }
+
+  if constexpr (I + 1 < N)
+    handleArg<Name, I + 1, N>(Args);
+}
+
+template<ConstexprString Name, typename... T>
+static void handleArgs(T &&...Args) {
+  if constexpr (sizeof...(T) > 0)
+    handleArg<Name, 0, sizeof...(T)>(std::make_tuple(Args...));
+}
+
+// This function will be used in each PipelineC function we need to wrap
+// For example:
+// rp_initialize(...) { return wrap<"rp_initialize">(_rp_initialize, ...); }
+template<ConstexprString Name, typename CalleeT, typename... ArgsT>
+static decltype(auto) wrap(CalleeT Callee, ArgsT... Args) {
+  using ReturnT = typename decltype(std::function{ Callee })::result_type;
+  if (TracingEnabled) {
+    std::lock_guard guard(TraceLock);
+    Tracer->functionPrelude(std::string_view(Name));
+    handleArgs<Name>(Args...);
+    Tracer->endArguments();
+    if constexpr (std::is_same_v<ReturnT, void>) {
+      Callee(std::forward<ArgsT>(Args)...);
+      Tracer->printVoid();
+    } else {
+      ReturnT Return = Callee(std::forward<ArgsT>(Args)...);
+      Tracer->printValue<std::remove_reference_t<ReturnT>>(Return);
```

Comment:
Why remove reference? Can't you let it auto deduce the template argument?

Reply:
Removed explicit template parameter